### DISCUSSION TITLE
Add rspamd milters

### DISCUSSION THREAD

**Discussion by Infern1** [msimerson/Mail-Toaster-6#530] [Ideas]
I disabled spamassassin and I know see mails come through where rspamd marks them as spam

Enable the function of rspamd to add headers.

https://rspamd.com/doc/modules/milter_headers.html

When I add this to `local.d/milter_headers.conf`

```
# local.d/milter_headers.conf:

# Options

# Add "extended Rspamd headers" (default false) (enables x-spamd-result, x-rspamd-server & x-rspamd-queue-id routines)
extended_spam_headers = true;

# List of headers to be enabled for authenticated users (default empty)
# authenticated_headers = ["authentication-results"];

# List of headers to be enabled for local IPs (default empty)
# local_headers = ["x-spamd-bar"];

# Set false to always add headers for local IPs (default true)
# skip_local = true;

# Set false to always add headers for authenticated users (default true)
# skip_authenticated = true;
#
use = ["x-spamd-bar", "x-spamd-result", "x-spam-status"];
```

The following header is added:
`X-Spam-Status: No, score=-0.50`

If this header is there our default mailfilter will act on it

```
if ( /X-Spam-Status: Yes/:h)
{
    if ( /X-Spam-Status: Yes, (hits|score)=([\d\.\-]+)\s/:h)
    {
...
```

**Comment by msimerson** [comment]
This suggestion interferes with the use of spamassassin and rspamd together. Better would be to update the [mailfilter](https://github.com/msimerson/Mail-Toaster-6/blob/master/qmail/filter.txt) to also check for the existing `X-Rspamd-Score` header, and maybe also the `X-Haraka-Karma` header.

**Reply by msimerson** [reply]
Better that using maildrop's mailfilter would be using sieve filters.

**Reply by Infern1** [reply]
> Better that using maildrop's mailfilter would be using sieve filters.

agree, but can we achieve the same functionality?
I am not a big fan of having two options for filtering. Maildrop done by qmailadmin and Sieve by others

**Reply by msimerson** [reply]
I would rather deprecate maildrop entirely and use sieve exclusively for this purpose, as it lets users (via sieve plugins for webmail) also create sieve scripts. (the way sqwebmail does for maildrop rules).

**Reply by Infern1** [reply]
> I would rather deprecate maildrop entirely and use sieve exclusively for this purpose, as it lets users (via sieve plugins for webmail) also create sieve scripts. (the way sqwebmail does for maildrop rules).

Yes I agree, most users I have use horde imp for filtering which exports to maildrop, the migration for me is easy because I tell horde imp to export to sieve instead.

I haven't tested sieve for a while, but is it fully working?

**Reply by msimerson** [reply]
I've had it running and enabled on my dovecot instance for years now. :-)

**Reply by msimerson** [reply]
Today I fiddled with this a little more. I made only this change:

````
# cat local.d/milter_headers.conf 

# Add headers: x-spamd-result, x-rspamd-server & x-rspamd-queue-id
extended_spam_headers = true;
````

Having done so, rspamd adds a `X-Spamd-Result` header which looks like this:

````
X-Spamd-Result: default: False [-1.00 / 15.00];
	DMARC_POLICY_ALLOW(-0.50)[tnpi.net,reject];
	R_DKIM_ALLOW(-0.20)[tnpi.net:s=mar2013];
	R_SPF_ALLOW(-0.20)[+mx];
	MIME_GOOD(-0.10)[text/plain];
	APPLE_MAILER_COMMON(0.00)[];
	RCVD_VIA_SMTP_AUTH(0.00)[];
	RCPT_COUNT_ONE(0.00)[1];
	MIME_TRACE(0.00)[0:+];
	ASN(0.00)[asn:7819, ipnet:66.128.48.0/20, country:US];
	MID_RHS_MATCH_FROM(0.00)[];
	RCVD_TLS_ALL(0.00)[];
	RECEIVED_SPAMHAUS_PBL(0.00)[97.113.150.69:received];
	FROM_EQ_ENVFROM(0.00)[];
	FROM_HAS_DN(0.00)[];
	ARC_NA(0.00)[];
	RCVD_COUNT_TWO(0.00)[2];
	TO_MATCH_ENVRCPT_ALL(0.00)[];
	TO_DN_NONE(0.00)[];
	DKIM_TRACE(0.00)[tnpi.net:+]
````

I then edited my sieve rules using snappymail's built-in sieve script editor as follows.

<img width="732" alt="Screenshot 2023-09-30 at 2 37 53 PM" src="https://github.com/msimerson/Mail-Toaster-6/assets/261635/5e0948d8-7781-4b2c-ace0-d0a014e6f4e1">

----
I have a similar rule for SpamAssassin's `X-Spam-Status` header. This embraces the legacy behavior of the old maildrop script and extends it for use with rspamd.
